src/cose/mac0.rs

```diff
@@ -0,0 +1,410 @@
+    MalformedTag(MacError),
+    #[error("tag is already present")]
+    AlreadyTagged,
+    #[error("error occurred when tagging COSE_Mac0: {0}")]
+    Tagging(MacError),
+    #[error("unable to set ClaimsSet: {0}")]
+    UnableToDeserializeIntoClaimsSet(CoseError),
+}
+
+/// Result with error type: [`Error`].
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Result for verification of a COSE_Mac0.
+#[derive(Debug)]
+pub enum VerificationResult {
+    Success,
+    Failure(String),
+    Error(Error),
+}
+
+impl VerificationResult {
+    /// Result of verification.
+    ///
+    /// `false` implies the signature is inauthentic or the verification algorithm encountered an
+    /// error.
+    pub fn is_success(&self) -> bool {
+        matches!(self, VerificationResult::Success)
+    }
+
+    /// Translate to a std::result::Result.
+    ///
+    /// Converts failure reasons and errors into a String.
+    pub fn into_result(self) -> Result<(), String> {
+        match self {
+            VerificationResult::Success => Ok(()),
+            VerificationResult::Failure(reason) => Err(reason),
+            VerificationResult::Error(e) => Err(format!("{}", e)),
+        }
+    }
+
+    /// Retrieve the error if the verification algorithm encountered an error.
+    pub fn into_error(self) -> Option<Error> {
+        match self {
+            VerificationResult::Error(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl PreparedCoseMac0 {
+    pub fn new(
+        builder: coset::CoseMac0Builder,
+        detached_payload: Option<&[u8]>,
+        aad: Option<&[u8]>,
+        tagged: bool,
+    ) -> Result<Self> {
+        let cose_mac0 = builder.build();
+
+        // Check if the payload is present and if it is attached or detached.
+        // Needs to be exclusively attached or detached.
+        let payload = match (cose_mac0.payload.as_ref(), detached_payload) {
+            (Some(_), Some(_)) => return Err(Error::DoublePayload),
+            (None, None) => return Err(Error::NoPayload),
+            (Some(payload), None) => payload.clone(),
+            (None, Some(payload)) => payload.to_vec(),
+        };
+        // Create the signature payload ot be used later on signing.
+        let tag_payload = mac_structure_data(
+            MacContext::CoseMac0,
+            cose_mac0.protected.clone(),
+            aad.unwrap_or_default(),
+            &payload,
+        );
+
+        Ok(Self {
+            cose_mac0: MaybeTagged::new(tagged, cose_mac0),
+            tag_payload,
+        })
+    }
+
+    /// Returns the signature payload that needs to be used to tag it.
+    pub fn signature_payload(&self) -> &[u8] {
+        &self.tag_payload
+    }
+
+    /// Finalize the COSE_Mac0 by adding the tag.
+    pub fn finalize(self, tag: Vec<u8>) -> MaybeTagged<CoseMac0> {
+        let mut cose_mac0 = self.cose_mac0;
+        cose_mac0.inner.tag = tag;
+        cose_mac0
+    }
+}
+
+impl MaybeTagged<CoseMac0> {
```

Comment:
impl for `MaybeTagged` wrapper isn't needed since the `Deref` implementations will allow this anyway.
```suggestion
impl CoseMac0 {
```

Reply:
`CoseMac0` is from `coset` crate and in that case getting error `Inherent "impl" cannot be defined for a type outside its containing crate [E0116]`

Reply:
Could be done with a trait, but it's not a big deal

Reply:
I understand, should I add it or leave it `impl MaybeTagged<CoseMac0> {` for now and maybe add the trait later?